hydromt/components/mesh.py

```diff
@@ -11,33 +11,370 @@
 import xugrid as xu
 from pyproj import CRS
 
-from hydromt import workflows
+from hydromt.components.base import ModelComponent
 from hydromt.gis.raster import GEO_MAP_COORD
-from hydromt.models.api import Model
+from hydromt.io.readers import read_nc
+from hydromt.models import Model
+from hydromt.workflows.mesh import (
+    create_mesh2d,
+    mesh2d_from_raster_reclass,
+    mesh2d_from_rasterdataset,
+    rename_mesh,
+)
 
-__all__ = ["MeshModel"]
-logger = logging.getLogger(__name__)
+logger = getLogger(__name__)
 
 
-class MeshMixin(object):
-    # placeholders
-    # We cannot initialize an empty xu.UgridDataArray
-    _API = {
-        "mesh": Union[xu.UgridDataArray, xu.UgridDataset],
-    }
+class MeshComponent(ModelComponent):
+    """ModelComponent class for mesh components."""
 
-    def __init__(self, *args, **kwargs) -> None:
-        super().__init__(*args, **kwargs)
-        self._mesh = None
+    def __init__(self, model: Model):
+        super().__init__(model=model)
+        self._data = None
 
-    ## general setup methods
-    def setup_mesh2d_from_rasterdataset(
+    def set(
         self,
+        data: Union[xu.UgridDataArray, xu.UgridDataset],
+        *,
+        name: Optional[str] = None,
+        grid_name: Optional[str] = None,
+        overwrite_grid: bool = False,
+    ) -> None:
+        """Add data to mesh.
+
+        All layers of mesh have identical spatial coordinates in Ugrid conventions.
+        Also updates self.region if grid_name is new or overwrite_grid is True.
+
+        Parameters
+        ----------
+        data: xugrid.UgridDataArray or xugrid.UgridDataset
+            new layer to add to mesh, should contain only one grid topology.
+        name: str, optional
+            Name of new object layer, this is used to overwrite the name of
+            a UgridDataArray.
+        grid_name: str, optional
+            Name of the mesh grid to add data to. If None, inferred from data.
+            Can be used for renaming the grid.
+        overwrite_grid: bool, optional
+            If True, overwrite the grid with the same name as the grid in self.mesh.
+        """
+        # Checks on data
+        data = _check_UGrid(data, name)
+
+        # Checks on grid topology
+        # TODO: check if we support setting multiple grids at once. For now just one
+        if len(data.ugrid.grids) > 1:
+            raise ValueError(
+                "set_mesh methods only supports adding data to one grid at a time."
+            )
+        if grid_name is None:
+            grid_name = data.ugrid.grid.name
+        elif grid_name != data.ugrid.grid.name:
+            data = rename_mesh(data, name=grid_name)
+
+        crs = self._add_mesh(
+            data=data, grid_name=grid_name, overwrite_grid=overwrite_grid
+        )
+
+        if crs:  # Restore crs
+            for grid in self.data.ugrid.grids:
+                grid.set_crs(crs)
```

Comment:
Nice that you did a new function, I would however also include updating crs in that function as well instead of here.